packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -0,0 +1,229 @@
+            bytes32 txHash = EIP712.hashPacked(
+                ctx.DOMAIN_SEPARATOR,
+                keccak256(
+                    abi.encode(
+                        WITHDRAWAL_TYPEHASH,
+                        withdrawal.owner,
+                        withdrawal.accountID,
+                        withdrawal.nonce,
+                        withdrawal.tokenID,
+                        withdrawal.amount,
+                        withdrawal.feeTokenID,
+                        withdrawal.fee,
+                        withdrawal.to,
+                        withdrawal.dataHash,
+                        withdrawal.minGas
+                    )
+                )
+            );
+
+            // Verify the signature if one is provided, otherwise fall back to an approved tx
+            if (auxData.signature.length > 0) {
+                require(txHash.verifySignature(withdrawal.owner, auxData.signature), "INVALID_SIGNATURE");
+            } else {
+                require(S.approvedTx[withdrawal.owner][txHash], "TX_NOT_APPROVED");
+                delete S.approvedTx[withdrawal.owner][txHash];
+            }
+        } else if (withdrawal.withdrawalType == 2 || withdrawal.withdrawalType == 3) {
+            // Forced withdrawals cannot make use of certain features because the
+            // necessary data is not authorized by the account owner.
+            // For protocol fee withdrawals, `owner` and `to` are both address(0).
+            require(withdrawal.owner == withdrawal.to, "INVALID_WITHDRAWAL_ADDRESS");
+
+            // Forced withdrawal fees are charged when the request is submitted.
+            require(withdrawal.fee == 0, "FEE_NOT_ZERO");
+
+            require(auxData.auxiliaryData.length == 0, "AUXILIARY_DATA_NOT_ALLOWED");
+
+            ExchangeData.ForcedWithdrawal storage forcedWithdrawal =
+                S.pendingForcedWithdrawals[withdrawal.accountID][withdrawal.tokenID];
+
+            if (forcedWithdrawal.timestamp != 0) {
+                if (withdrawal.withdrawalType == 2) {
+                    require(withdrawal.owner == forcedWithdrawal.owner, "INCONSISENT_OWNER");
+                } else { //withdrawal.withdrawalType == 3
+                    require(withdrawal.owner != forcedWithdrawal.owner, "INCONSISENT_OWNER");
+                    require(withdrawal.amount == 0, "UNAUTHORIZED_WITHDRAWAL");
+                }
+
+                // Get the fee
+                feeETH = forcedWithdrawal.fee;
+
+                // delete the withdrawal request and free a slot
+                delete S.pendingForcedWithdrawals[withdrawal.accountID][withdrawal.tokenID];
+                S.numPendingForcedTransactions--;
+
+                emit ForcedWithdrawalProcessed(
+                    withdrawal.accountID,
+                    withdrawal.tokenID,
+                    withdrawal.amount
+                );
+            } else {
+                // Allow the owner to submit full withdrawals without authorization
```

Comment:
For this case, we should check it in circuit for such off-chain request. Need to double check them on circuit.

Reply:
Can you clarify what you mean with that? I'm not sure I understand. In this case there would be no offchain request that's signed by the user, and the circuit doesn't know directly onchain what's going on, so I'm not sure what you'd want to be checked in the circuit as well.

Reply:
> Can you clarify what you mean with that? I'm not sure I understand. In this case there would be no offchain request that's signed by the user, and the circuit doesn't know directly onchain what's going on, so I'm not sure what you'd want to be checked in the circuit as well.

I mean withdraw protocol fee.

Reply:
I'm afraid I still don't follow. What do we have to check in the circuits for withdrawing from the protocol fee account? This account is always at `accountID == 0` and the operator cannot lie about the accountID put onchain because the accountID used in the circuit is put into the DA to be used here, so we can be sure it is correct and matches the one used in the circuit.